src/crystal/system/unix/file.cr

```diff
@@ -0,0 +1,236 @@
+require "c/sys/file"
+
+# :nodoc:
+module Crystal::System
+  module File
+    extend self
+
+    def open(filename, mode, perm)
+      oflag = open_flag(mode) | LibC::O_CLOEXEC
+
+      fd = LibC.open(filename.check_no_null_byte, oflag, perm)
+      if fd < 0
+        raise Errno.new("Error opening file '#{filename}' with mode '#{mode}'")
+      end
+      fd
+    end
+
+    private def open_flag(mode)
+      if mode.size == 0
+        raise "Invalid access mode #{mode}"
+      end
+
+      m = 0
+      o = 0
+      case mode[0]
+      when 'r'
+        m = LibC::O_RDONLY
+      when 'w'
+        m = LibC::O_WRONLY
+        o = LibC::O_CREAT | LibC::O_TRUNC
+      when 'a'
+        m = LibC::O_WRONLY
+        o = LibC::O_CREAT | LibC::O_APPEND
+      else
+        raise "Invalid access mode #{mode}"
+      end
+
+      case mode.size
+      when 1
+        # Nothing
+      when 2
+        case mode[1]
+        when '+'
+          m = LibC::O_RDWR
+        when 'b'
+          # Nothing
+        else
+          raise "Invalid access mode #{mode}"
+        end
+      else
+        raise "Invalid access mode #{mode}"
```

Comment:
Doesn't support a mode like `rb+`. Also, `#{mode.inspect}` would make it clearer in the output what comes from the user and what's the error message itself

Reply:
This is directly copied from the old `file.cr` if it has issues, it's out of the scope of this PR.

Reply:
Fair enough